### DISCUSSION TITLE
Technology selection

### DISCUSSION THREAD

**Discussion by amis92** [BSData/phalanx#3] [General]
## Intro

Having user stories is all fine and dandy - we decided *what* we want to do. Now, we also need to decide *how* and select technologies and tools we'll use.

## How we do it

I suggest we discuss separate "topics" via replies to "topic comments". Want new topic, start a new "thread"/comment.

**Comment by amis92** [comment]
# Application model

What do we do? Vote! (add a reaction

- 🚀 Do we challenge the native app world (iOS, Android, Windows, macOS, linux)

- 🎉 Do we go to the *web*-way and do an offline-capable, fully client-side web app hosted statically on GitHub Pages

*or maybe you have another idea - share it!*

**Reply by rweyrauch** [reply]
The web model is the easiest.  Only need a single app.  Distribution is simple.  Give up platform specific look and feel.

**Reply by amis92** [reply]
Differences I can think of tgat might matter to us:

- native has access to file storage
- native has rich sharing API
- native requires an identity to own and publish on appstores
- native requires cross targeting for "universal"/xplat frameworks (e.g. Xamarin) OR separate apps

- web requires no installation
- web means single app
- web requires RWD to scale and adapt behavior across form factors and inout methods
- web can be statically hosted on e.g. Github Pages so free publishing
- web has browser constraints on web usage (e.g. CORS)
- web has no adaptation to OS look, feel and behaviors
- web works everywhere (like new OSes)

**Reply by amis92** [reply]
Voting time!

**Reply by cartag** [reply]
How will we handle doing data files separate from the roster editor if it's a web app? For saving rosters in a webapp, would it require sign in? Who would do authentication for it?

**Reply by amis92** [reply]
Valid questions. I don't have definite answer, but I'll try:
1. It'll be a client-side webapp, static, offline-capable PWA. Essentially, it's an app in a browser. Server can be GitHub Pages hosting static resource files. So, data files are of any kind will be downloaded to the client's browser the same as in BScribe, via AppSpot/Gallery repo list.
2. Saving rosters on the device wouldn't require signin, but having them somehow sync'd across devices/saved in the cloud - definitely would. We can integrate with Google Account or Microsoft Account authentication and ask user for permission to their GDrive/OneDrive. I believe it's a future feature, to somehow enable saving rosters in the cloud.
For starters though, we can just save in the browser and allow user to export to a file (would be "downloaded" in the browser).

**Reply by TrentBartlem** [reply]
baked in support for cloud sync is a good idea. MVP is client-side storage - still need it for any edits that are done offline and before save

**Reply by CloverFox** [reply]
I think client side webapp will be a great first step/PoC, things like sharing and platform specific theming ect would be after MVP anyway

**Reply by Manethak** [reply]
I second the web app. It alleviates the issues of specific ui for specific apps or environments and will get us off the ground faster. Once the back end is done the front end can change to meet the needs

**Comment by amis92** [comment]
# Languages and Frameworks

This is cross-cutting with **Application model**. So let's also hear what our community's *forte* is. We can't write an xplat native app in `C++` ... or can we.

1. Reply with how you can do webapps and/or native apps (what framework/language).
2. Share an opinion what could be more favorable and why.

**Reply by rweyrauch** [reply]
Language is largely dependent on the choice of target platform.   A mashup of Wasm and JavaScript would suffice for the web app.   The Wasm could be generated from C# or C++.

**Reply by amis92** [reply]
@rweyrauch these are intertwined. We need to see what our people can do, and then decide what we can afford, with possibly the preferred appmodel being discarded due to no implementers.

**Reply by amis92** [reply]
I'm good at C#/.NET/Blazor and PowerShell, GitHub Actions and git

Can do some JavaScript and TypeScript

**Reply by TrentBartlem** [reply]
I'm okay at Java, kotlin et al. Have done a bit of JS

**Reply by rweyrauch** [reply]
In my day job I'm a c++/c# dev.  I tinker with Typescript in my free time.

**Reply by CloverFox** [reply]
Java dev by trade, dabbled in client side JS,
I think due to the webapp/mobile centric, although its none of our primary languages, i think one of the JS frameworks might be the way to go

**Reply by Manethak** [reply]
Personally I have a lot of experience with c#, some c++, xamarin, Java, Android, objective c, blazor.

I teach c++ and c#

**Reply by Thairne** [reply]
Java/Javascript by day, Typescript in the dark hours when I wail in dispair. Going to get better in that due to switch to TS in company.

**Reply by nicolasconnault** [reply]
JS/TS with some experience with Java and Flutter

**Reply by mwadams** [reply]
If this gets a bit of momentum, I may be able to help out a bit. I can contribute C#/JS/TS and have a lot of experience with API and service design, and the edge/cloud problem space.
